### PR TITLE
H7 FDCAN driver.

### PR DESCRIPTION
**Pull request by iabdalkader**
* This is submitted for **review and edits** related to issue #4774

* This driver is implemented in fdcan.c but it defines pyb_can_type, so from a user's perspective it's same as can.c (ex. from pyb import CAN works the same). fdcan.c is not compiled unless HW_FDCAN_TX/RX is defined and it shares can.h with can.c, so there's no need for an additional fdcan_init0()/deinit etc...

* A new GPIO function (FDCAN) had to be added, otherwise the AF_FN_CAN couldn't be found in runtime, or the build would fail if AF_FN_FDCAN is used.

* The "Message RAM" for FDCAN 1 and 2 is shared and a memory offset can be configured for CAN controller memory. I split the memory in half between the two CANs by using memory_size/2 as offset for CAN2. This offset simplifies managing things like filter indices (which always start from 0->64 for either CAN controller), see RM0433 56.3.2

* The setfilter function only supports the RANGE mode. This allows a range of IDs, ex. `can.setfilter(index, fifo, (id_start, id_end)`. Filters support other modes like dual IDs, bit masking ID and other configuration modes, see RM0433 56.3.19.

* This driver has been tested in loopback mode and with a two node transceiver-less network. However, it could use a lot of improvements especially with filters, interrupts and error handling, rxcallback etc...

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks very much for this.  After #4784 I will look at this in detail.

**Comment by iabdalkader** [comment]
I'm going to continue working on this now, and I have CAN transceivers to test with. I will rebase to fix conflicts, but I need to know if adding a new driver like this is okay or do you prefer to support them both in the same driver ?

**Comment by iabdalkader** [comment]
Update: I tested this driver with transceivers and it's working fine, also fixed a bug with any() and filter acceptance and added support for the other filter types.

**Comment by iabdalkader** [comment]
@dpgeorge  I really could use some help/hints here, as I want to merge this in our next release (which is over due) and I don't want to diverge too much from upstream.

**Comment by dpgeorge** [comment]
Ok, I'll look into it.

**Comment by dpgeorge** [comment]
> but I need to know if adding a new driver like this is okay or do you prefer to support them both in the same driver ?

I can understand why you added fdcan.c, because it's a bit different to the existing can.c.  But I do thing there are enough similarities (especially with the Python-level API) that these files can be merged and can.c can support both CAN and FDCAN peripherals (similar to how sdcard.c also supports MMC).

I simplified this PR in #5096, trying to reduce the diff.  There are less files changed (AF and pin stuff remain the same, with a small #define to make it work), and the diff between can.c and fdcan.c is smaller by about 10% (minor changes, nothing functional).

An alternative to merging the code together in can.c would be to have 3 files:
- can.c with low-level helper functions for CAN periph
- fdcan.c with low-level helper functions for FDCAN periph
- pyb_can.c with Python-level wrapper functions, using low-level helpers from above files

Doing that probably requires a similar effort to merging can.c with fdcan.c.  And it's probably a bit cleaner to do it this way.

**Comment by iabdalkader** [comment]
> I simplified this PR in #5096, trying to reduce the diff. There are less files changed (AF and pin stuff remain the same, with a small #define to make it work),

This is much better, but I think the test should be for MICROPY_HW_ENABLE_FDCAN because other (future) micros may have FDCAN as well. Maybe define in mpconfigboard_common.h for STM32H7 ?

> An alternative to merging the code together in can.c would be to have 3 files:
> 
> * can.c with low-level helper functions for CAN periph
> * fdcan.c with low-level helper functions for FDCAN periph
> * pyb_can.c with Python-level wrapper functions, using low-level helpers from above files

I like this approach more, but shouldn't it be called machine_can.c now instead ?

**Comment by dpgeorge** [comment]
> but I think the test should be for MICROPY_HW_ENABLE_FDCAN because other (future) micros may have FDCAN as well. Maybe define in mpconfigboard_common.h for STM32H7 ?

That sounds good.

> I like this approach more, but shouldn't it be called machine_can.c now instead ?

Ok, let's go for this approach then, since it's more sustainable for the future, and matches how other peripherals are evolving.

But it should be `pyb_can.c` because this class is very specific to the stm32 port.  One day we can define `machine.CAN`, which would also work on the esp32 and other targets, and then utilize `stm32/can.c` and `stm32/fdcan.c` to implement it on stm32.

**Comment by dpgeorge** [comment]
See #5100 for an initial attempt at splitting out Python bindings to `pyb_can.c` (no FDCAN support though).

**Comment by iabdalkader** [comment]
@dpgeorge I pulled changes from your PRs into this branch, and I'm adding FDCAN support... Will push more commits later.

**Comment by iabdalkader** [comment]
@dpgeorge It's done.

**Comment by dpgeorge** [comment]
>  It's done.

Ok, thanks.  I see that it could probably be factored a bit more, more low-level stuff in can.c/fdcan.c, but that doesn't have to be done now.

I'll review this fully as soon as I can.

**Comment by iabdalkader** [comment]
> I see that it could probably be factored a bit more, more low-level stuff in can.c/fdcan.c, but that doesn't have to be done now.

Yes I couldn't agree more, especially pyb_can_init_helper,  pyb_can_setfilter and pyb_can_send, but I already changed too much code in this PR.

Note I tested sending, receiving and setfilter on F7 and H7, it works fine but one minor issue with setfilter it uses different constants for CAN and FDCAN. I'm not sure what would be the best way to fix this. Anyway, I'm going to merge this into our branch please let me know if it needs more work.

**Comment by dpgeorge** [comment]
I had more of a look at this PR and it looks fine for an initial working version.

> setfilter it uses different constants for CAN and FDCAN

Yeah I looked at the datasheet and indeed the FDCAN hardware is different with the way it does filtering:
- RANGE mode: could be mapped to MASK32 but not in a fully general way
- DUAL mode: could be mapped to LIST32, just a list of 2 allowed id's
- MASK mode: could be mapped to MASK32 but probably not in a general way as the masking behaviour is slightly different to the original CAN peripheral

Well, the aim of pyb-modules is to provide a low-level interface to the hardware, and it's not possible to do that if the underlying hardware differs among MCUs.  Since we don't want to restrict too much how the underlying FDCAN is used I think it's fair enough to allow the config of the filters to differ here.  At least the signature of `setfilter()` will be the same for both versions, it's just the constants (LIST16, MASK16 vs RANGE, DUAL, MASK) will differ.

In summary I'm happy to have different constants and setfilter behaviour.

**Comment by iabdalkader** [comment]
> I think it's fair enough to allow the config of the filters to differ here. At least the signature of `setfilter()` will be the same for both versions, it's just the constants (LIST16, MASK16 vs RANGE, DUAL, MASK) will differ.
> 
> In summary I'm happy to have different constants and setfilter behaviour.

Then it should be good to merge now I think, otherwise let me know if there's something else that needs to be done in this PR.

**Comment by dpgeorge** [comment]
Ok, this driver was merged in f7a07b36057d61264958e3577f7e4555e77d520c, support for FDCAN on NUCLEO_H743ZI merged in 4f78ba3cf4db499e863547a05088012b87543cbb

**Comment by dpgeorge** [comment]
@iabdalkader thanks for the effort here.  Feel free to follow up with any fixes as you find them.
